src/test/java/net/minecraftforge/debug/misc/DataPackRegistriesTest.java

```diff
@@ -0,0 +1,228 @@
+import net.minecraftforge.forge.event.lifecycle.GatherDataEvent;
+import net.minecraftforge.registries.DeferredRegister;
+import net.minecraftforge.registries.ForgeRegistryEntry;
+import net.minecraftforge.registries.RegistryBuilder;
+import net.minecraftforge.registries.RegistryObject;
+
+/**
+ * <p>This test class shows an example of how to register unsynced and synced datapack registries,
+ * and how to use a dataprovider to generate json files for them.
+ * It also verifies that data and tags are loaded and synced correctly.
+ * Data is loaded from the following folders in test resources:</p>
+ * <ul>
+ * <li><code>data/data_pack_registries_test/data_pack_registries_test/unsyncable/test.json</code></li>
+ * <li><code>data/data_pack_registries_test/tags/data_pack_registries_test/unsyncable/test.json</code></li>
+ * <li><code>data/data_pack_registries_test/data_pack_registries_test/syncable/test.json</code></li>
+ * <li><code>data/data_pack_registries_test/tags/data_pack_registries_test/syncable/test.json</code></li>
+ * </ul>
+ */
+@Mod(DataPackRegistriesTest.MODID)
+public class DataPackRegistriesTest
+{
+    private static final boolean ENABLED = true;
+    public static final String MODID = "data_pack_registries_test";
+    private static final Logger LOGGER = LogUtils.getLogger();
+    private static final ResourceLocation TEST_RL = new ResourceLocation(MODID, "test");
+    
+    private final RegistryObject<Unsyncable> datagenTestObject;
+
+    public DataPackRegistriesTest()
+    {
+        if (ENABLED)
+        {
+            final IEventBus modBus = FMLJavaModLoadingContext.get().getModEventBus();
+            final IEventBus forgeBus = MinecraftForge.EVENT_BUS;
+            
+            // Deferred Registers can be created for datapack registries in static init or mod constructor.
+            // (We'll do it in mod constructor, as when doing the ENABLED check it's less verbose than static init.)
+            // As with static registries, any mod can make a Deferred Register for a given datapack registry,
+            // but only one mod can register the internal registry with makeRegistry.
+            final DeferredRegister<Unsyncable> unsyncables = DeferredRegister.create(Unsyncable.REGISTRY_KEY, MODID);
+            final DeferredRegister<Syncable> syncables = DeferredRegister.create(Syncable.REGISTRY_KEY, MODID);
+            
+            // RegistryBuilder#dataPackRegistry marks the registry as a datapack registry rather than a static registry.
+            unsyncables.makeRegistry(Unsyncable.class,
+                () -> new RegistryBuilder<Unsyncable>().disableSaving().dataPackRegistry(Unsyncable.DIRECT_CODEC));
+            // The overload of #dataPackRegistry that takes a second codec marks the datapack registry as syncable.
+            syncables.makeRegistry(Syncable.class,
+                () -> new RegistryBuilder<Syncable>().disableSaving().dataPackRegistry(Syncable.DIRECT_CODEC, Syncable.DIRECT_CODEC));
```

Comment:
See vanilla usages. Stuff that is only ever needed/used/processed by the server is just wasted networking data if sent through the main codec over networking.

Reply:
There also exist codecs that don't work at all when used for networking because they're hardcoded to be used by RegistryOps and the netcode doesn't use RegistryOps.